#! /bin/sh /usr/share/dpatch/dpatch-run

## 08_security_CAN-2005-0071.dpatch by Thomas Schmidt <tschmidt@debian.org>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Fixes CAN-2005-0071 (It is not possible to overwrite files with 
## DP: the GRAB-Command anymore)

@DPATCH@
diff -urNad vdr-1.2.6/dvbdevice.c /tmp/dpep.80S4AN/vdr-1.2.6/dvbdevice.c
--- vdr-1.2.6/dvbdevice.c	2004-04-22 12:48:31.000000000 +0200
+++ /tmp/dpep.80S4AN/vdr-1.2.6/dvbdevice.c	2005-01-16 21:51:24.000000000 +0100
@@ -505,8 +505,10 @@
               Quality = 255; //XXX is this 'best'???
 
            isyslog("grabbing to %s (%s %d %d %d)", FileName, Jpeg ? "JPEG" : "PNM", Quality, vm.width, vm.height);
-           FILE *f = fopen(FileName, "wb");
-           if (f) {
+	   int fd = open(FileName, O_CREAT | O_EXCL | O_TRUNC | O_RDWR, 00640);
+           if (fd > -1) {
+            FILE *f = fdopen(fd, "wb");
+            if (f) {
               if (Jpeg) {
                  // write JPEG file:
                  struct jpeg_compress_struct cinfo;
@@ -540,7 +542,8 @@
                     }
                  }
               fclose(f);
-              }
+	    }
+	   }
            else {
               LOG_ERROR_STR(FileName);
               result |= 1;
